docs: differences between lb3 and lb4 cli

emonddr opened the pull request:
Describe the similarities and differences between
the LB3 and LB4 command line interfaces

Related to : https://github.com/strongloop/loopback-next/issues/3953 

## Viewing this document
The best way to view this document is to:
- clone this repository 
- switch to the `dremond_cli_migration_doc` branch
- npm install
- npm run build
- npm run docs:prepare
- npm run docs:start

This will start a document server on your machine.
Navigate to `Navigation > CLI` section of LoopBack 4 documentation.
<!--
Please provide a high-level description of the changes made by your pull request.

Include references to all related GitHub issues and other pull requests, for example:

Fixes #123
Implements #254
See also #23
-->

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [ ] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

emonddr commented:
I am providing screenshots of the rendered document below for convenience purposes:

![image](https://user-images.githubusercontent.com/6864736/79626250-bda45980-80fc-11ea-8b64-a1ab1619194b.png)

![image](https://user-images.githubusercontent.com/6864736/79626266-d3198380-80fc-11ea-9dcd-853b861b15b1.png)

![image](https://user-images.githubusercontent.com/6864736/79626279-e593bd00-80fc-11ea-948a-53ea59fa2ad6.png)

raymondfeng commented on docs/site/migration/cli.md:
```diff
@@ -10,8 +10,23 @@ permalink: /doc/en/lb4/migration-cli.html
-following GitHub issue:
-[loopback-next#3953](https://github.com/strongloop/loopback-next/issues/3953)
-" %}
+The commandline interfaces of [LoopBack 3](https://loopback.io/doc/en/lb3/Command-line-tools.html) and [LoopBack 4](../Command-line-interface.md) have some similarities, but also some differences.
```
`commandline` -> `command line`

dhmlau commented on docs/site/migration/tables/lb3-lb4-command-mappings.html:
```diff
@@ -0,0 +1,83 @@
+
+      <td>
+        In LB3, create a fully functioning application using an API defined using swagger 2.0.
+        In LB4, create a fully functioning application using an OpenAPI 3.0 specification
```
i think `lb4 openapi` supports both Swagger 2.0 and OpenAPI 3.0.

dhmlau commented on docs/site/migration/tables/lb3-only-commands.html:
```diff
@@ -0,0 +1,104 @@
+      </td>
+      <td> Create a new application destined in Bluemix
+      </td>
+      <td> None
```
Does Appsody help to create LB4 app that is ready to deploy to IBM Cloud? 
If not, it might be useful to point it to https://loopback.io/doc/en/lb4/Deploying-to-IBM-Cloud.html.

dhmlau commented on docs/site/migration/tables/lb3-only-commands.html:
```diff
@@ -0,0 +1,104 @@
+      <td>
+        Creates a skeleton boot script in server/boot directory. Allows for customized intialization of anything during boot phase
+      </td>
+      <td> Create a <a href="https://loopback.io/doc/en/lb4/Booting-an-Application.html#custom-booters">custom booter</a> for a new artifact, or create a <a href="https://loopback.io/doc/en/lb4/Life-cycle.html#overview">lifecycle observer</a> to run custom code during different phases of the application's lifecycle including startup.
```
I think the boot script in LB3 is more similar to the lifecycle observer, and not too much about booter.  Perhaps we just need to mention the observer?

dhmlau commented on docs/site/migration/tables/lb3-only-commands.html:
```diff
@@ -0,0 +1,104 @@
+        <code>lb soap</code>
+      </td>
+      <td>
+        Create a SOAP client based on a soap service's WSDL file
```
`soap service` -> `SOAP service`

dhmlau commented on docs/site/migration/tables/lb4-only-commands.html:
```diff
@@ -0,0 +1,113 @@
+    <tr>
+      <th width="200">Commands</th>
+      <th>Description</th>
+    </tr>
```
Just a thought. It might help to have one more column about why there's no equivalent LB3 command (e.g. lb4 repository and lb4 controller), or the similar concept in LB3 (e.g. interceptor).

emonddr commented:
Addressed comments:

![image](https://user-images.githubusercontent.com/6864736/79915879-c98e6500-83f5-11ea-8909-bb9b0adff73c.png)

![image](https://user-images.githubusercontent.com/6864736/79915913-d743ea80-83f5-11ea-9d5f-495ae0279bee.png)

![image](https://user-images.githubusercontent.com/6864736/79915957-e9258d80-83f5-11ea-8979-754d50c27a21.png)

nabdelgadir commented on docs/site/migration/tables/lb4-only-commands.html:
```diff
@@ -0,0 +1,150 @@
+      </td>
+      <td> Create a repository for a model
+      </td>
+      <td> A new concept in LB4. A model's repository defines the CRUD operations. In LB3, the model defined the CRUD operations.
```
Just wondering whether it would be worth it to add links to the LB4 pages, like a link to https://loopback.io/doc/en/lb4/Repositories.html, so people can learn more about it there?

emonddr replied:
@nabdelgadir , I will add one for controllers as well.

jannyHou commented on docs/site/migration/tables/lb3-only-commands.html:
```diff
@@ -0,0 +1,104 @@
+      <td>
+        Adds a new remote method to a model
+      </td>
+      <td> Open the model's controller file and add a new method manually. Depending on the purpose of this new controller method, you may also need to a new method to the model's repository.
```
the remote method command actually generates the strong-remoting style method definitions that can be transferred to openapi spec and prints out the dummy function, then you need to manually copy the dummy function to a model's js file. It doesn't write a function directly into the js file.
Could you provide link here https://loopback.io/doc/en/lb4/Controllers.html#writing-controller-methods and briefly mention "decorate the controller function with REST decorators"?

emonddr replied:
![image](https://user-images.githubusercontent.com/6864736/80012817-e3828300-849b-11ea-9a80-ec6e610164a6.png)

Ok, so it adds it to the application, and not the model.

emonddr replied:
> Could you provide link here https://loopback.io/doc/en/lb4/Controllers.html#writing-controller-methods and briefly mention "decorate the controller function with REST decorators"? 

Good idea.

jannyHou commented on docs/site/migration/tables/lb3-only-commands.html:
```diff
@@ -0,0 +1,104 @@
+      <td>
+        Creates a skeleton boot script in server/boot directory. Allows for customized intialization of anything during boot phase
+      </td>
+      <td> Create a <a href="https://loopback.io/doc/en/lb4/Life-cycle.html#overview">lifecycle observer</a> to run custom code during different phases of the application's lifecycle; including startup.
```
probably mention the migration guide for boot script in https://loopback.io/doc/en/lb4/migration-boot-scripts.html

jannyHou commented on docs/site/migration/tables/lb4-only-commands.html:
```diff
@@ -0,0 +1,150 @@
+      </td>
+      <td> Create a service
+      </td>
+      <td> A new concept in LB4.
```
better to add a link https://loopback.io/doc/en/lb4/Services.html

jannyHou commented on docs/site/migration/tables/lb4-only-commands.html:
```diff
@@ -0,0 +1,150 @@
+      </td>
+      <td> Create an interceptor
+      </td>
+      <td> A new concept in LB4.
```
same here, https://loopback.io/doc/en/lb4/Interceptors.html

jannyHou commented on docs/site/migration/tables/lb4-only-commands.html:
```diff
@@ -0,0 +1,150 @@
+      </td>
+      <td> Create a lifecycle observer
+      </td>
+      <td> A new concept in LB4.
```
same here: https://loopback.io/doc/en/lb4/Life-cycle.html

dhmlau commented on docs/site/migration/tables/lb3-lb4-command-mappings.html:
```diff
@@ -0,0 +1,83 @@
+
+    <tr>
+      <td>
+        <code>lb</code>
```
actually `lb` and `lb app` will create a new app

dhmlau commented on docs/site/migration/tables/lb3-lb4-command-mappings.html:
```diff
@@ -0,0 +1,83 @@
+        <code>lb</code>
+      </td>
+      <td>
+        <code>lb4 app</code>
```
similarly for `lb4` and `lb4 app`

dhmlau commented on docs/site/migration/tables/lb3-only-commands.html:
```diff
@@ -0,0 +1,104 @@
+      <td>
+        Create a SOAP client based on a SOAP service's WSDL file
+      </td>
+      <td> Manually create an application that calls the SOAP service through the loopback-connector-soap connector.
```
I'm not sure if we have an equivalent or workaround for that. @raymondfeng @bajtos might be the better person to comment on that.

emonddr replied:
Right now, there is no equivalent to `lb soap` (from LB3) in LB4. And all the documentation and code samples for interacting with web services are written for LB3. So, currently, I cannot even point users to a LB4 code snippet for any soap related stuff. This is why I mentioned this a few weeks ago in an estimation meeting. @bajtos , @raymondfeng , perhaps we can mention a backlog task url here in the meantime; until we have an equivalent command or LB4 code snippets documented?

dhmlau replied:
I think the text in the table might refer to https://loopback.io/doc/en/lb4/Calling-other-APIs-and-web-services.html, which is not quite the same as `lb soap`. This link shows the steps on how to connect to a SOAP service but it does not generate the application based on a WSDL. Perhaps before we have such functionality, we can mention this?

emonddr commented:
Latest render screencaps:

![image](https://user-images.githubusercontent.com/6864736/80020684-6b21bf00-84a7-11ea-8366-b14a8f7695fc.png)


![image](https://user-images.githubusercontent.com/6864736/80020737-7d9bf880-84a7-11ea-8e98-85a195228cf9.png)

![image](https://user-images.githubusercontent.com/6864736/80020799-986e6d00-84a7-11ea-9aae-964ccac2327d.png)

nabdelgadir commented on docs/site/migration/tables/lb4-only-commands.html:
```diff
@@ -0,0 +1,150 @@
+      </td>
+      <td> Create a new extension
+      </td>
+      <td> A new utility command in LB4. LB3 modules are stored in separate github repositories whereas LB4 modules are stored in a single repository ( a monorepo). This LB4 command allows a contributor to easily add a new extension subdirectory to the monorepo.
```
nitpick: remove space after bracket `( a monorepo`

nabdelgadir commented on docs/site/migration/tables/lb3-lb4-command-mappings.html:
```diff
@@ -0,0 +1,83 @@
+      <td>
+        <code>lb4 model</code>
+      </td>
+      <td> Create a new model. In LB3, this would create the model, but you would automatically get all the model methods and remote methods. In LB4, this command only creates the model; you must also create a repository to handle the CRUD operations of your model, and you must create a controller to define your model's REST API endpoints. Alternatively, you can use 'lb4 rest-crud' to create a co ...
```
I think similar to the commands in the `Commands` column, it can be `<code>lb4 rest-crud</code>`

emonddr commented:
Latest render of document:

![image](https://user-images.githubusercontent.com/6864736/80247918-ea94c700-863c-11ea-8bf8-26a52193193f.png)

![image](https://user-images.githubusercontent.com/6864736/80247957-fda79700-863c-11ea-9774-a48b6cfd089d.png)

![image](https://user-images.githubusercontent.com/6864736/80248002-16b04800-863d-11ea-918c-6175ccdf1346.png)
